routers/web/repo/issue.go

```diff
@@ -1230,45 +1230,78 @@ func NewIssuePost(ctx *context.Context) {
+	roleDescriptor := issues_model.RoleDescriptor{}
+
 	if hasOriginalAuthor {
-		return issues_model.RoleDescriptorNone, nil
+		return roleDescriptor, nil
 	}
 
 	perm, err := access_model.GetUserRepoPermission(ctx, repo, poster)
 	if err != nil {
-		return issues_model.RoleDescriptorNone, err
+		return roleDescriptor, err
 	}
 
-	// By default the poster has no roles on the comment.
-	roleDescriptor := issues_model.RoleDescriptorNone
+	// If the poster is the actual poster of the issue, enable Poster role.
+	roleDescriptor.IsPoster = issue.IsPoster(poster.ID)
 
 	// Check if the poster is owner of the repo.
 	if perm.IsOwner() {
 		// If the poster isn't a admin, enable the owner role.
 		if !poster.IsAdmin {
-			roleDescriptor = roleDescriptor.WithRole(issues_model.RoleDescriptorOwner)
-		} else {
+			roleDescriptor.Role = issues_model.RoleDescriptorOwner
+			return roleDescriptor, nil
+		}
 
-			// Otherwise check if poster is the real repo admin.
-			ok, err := access_model.IsUserRealRepoAdmin(repo, poster)
-			if err != nil {
-				return issues_model.RoleDescriptorNone, err
-			}
-			if ok {
-				roleDescriptor = roleDescriptor.WithRole(issues_model.RoleDescriptorOwner)
-			}
+		// Otherwise check if poster is the real repo admin.
+		ok, err := access_model.IsUserRealRepoAdmin(repo, poster)
+		if err != nil {
+			return roleDescriptor, err
+		}
+		if ok {
+			roleDescriptor.Role = issues_model.RoleDescriptorOwner
+			return roleDescriptor, nil
 		}
 	}
 
-	// Is the poster can write issues or pulls to the repo, enable the Writer role.
-	// Only enable this if the poster doesn't have the owner role already.
-	if !roleDescriptor.HasRole("Owner") && perm.CanWriteIssuesOrPulls(issue.IsPull) {
-		roleDescriptor = roleDescriptor.WithRole(issues_model.RoleDescriptorWriter)
+	// If repo is organization, check Member role
+	if err := repo.LoadOwner(ctx); err != nil {
+		return roleDescriptor, err
+	}
+	if repo.Owner.IsOrganization() {
+		if isMember, err := organization.IsOrganizationMember(ctx, repo.Owner.ID, poster.ID); err != nil {
+			return roleDescriptor, err
+		} else if isMember {
+			roleDescriptor.Role = issues_model.RoleDescriptorMember
+			return roleDescriptor, nil
+		}
 	}
 
-	// If the poster is the actual poster of the issue, enable Poster role.
-	if issue.IsPoster(poster.ID) {
-		roleDescriptor = roleDescriptor.WithRole(issues_model.RoleDescriptorPoster)
+	// If the poster is the collaborator of the repo
+	if isCollaborator, err := repo_model.IsCollaborator(ctx, repo.ID, poster.ID); err != nil {
+		return roleDescriptor, err
+	} else if isCollaborator {
+		roleDescriptor.Role = issues_model.RoleDescriptorCollaborator
+		return roleDescriptor, nil
+	}
+
+	// If the poster is the contributor of the repo
+	searchOpt := &issue_indexer.SearchOptions{
+		Paginator: &db.ListOptions{
+			Page:     1,
+			PageSize: 1,
+		},
+		RepoIDs:  []int64{repo.ID},
+		IsClosed: util.OptionalBoolTrue,
```

Comment:
IsMerged? Maybe we haven't have that option.

Reply:
Don't have this option.

Reply:
@wolfogre 
any plan to add this option?

Reply:
I think some inaccurate tags here won't affect end user's experience. The worse case is a "first-time contributor" is considers as "contributor".

Reply:
see 399f31de1525fd49a9ed9ca1a171d2f4ba1fae48 , sorry looks like I pushed wrong commits. Maybe you can have a force push after pull.